### PR TITLE
feat(api):finished API GraphQL

### PR DESCRIPTION
**Pull request by aitorarias**
Attach pull request with the complete exercise. 

P.S: still can't assign to @claradopico

### CODE REVIEW

**Comment by davidbarral** [comment]
Dont' worry, we'll pair review. Thanks.

File: index.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+const users = require("./data/users");
+
+const server = new ApolloServer({
+  typeDefs,
+  resolvers,
+  context: async ({ req }) => {
+    const auth = req ? req.headers.authorization : null;
+    if (auth && auth.startsWith("bearer ")) {
+      const token = auth.substring(7);
+      const verifyToken = jwt.verify(token, process.env.JWT_SECRET);
```
Comment (davidbarral) [inline index.js:17]: I guess you assigned this var to get the id from the token, but you are not using it anywhere.

File: index.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+      const currentUser = await users.find(({ id }) => id);
```
Comment (davidbarral) [inline index.js:18]: This will always find the first user, doesn't it?

Comment (davidbarral) [reply index.js:18]: Maybe we need to take the id from the token and find the correct user, and check if the user exists (and raise an error if it doesn't).

File: index.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+const jwt = require("jsonwebtoken");
+const typeDefs = require("./graphql/schema");
+const resolvers = require("./graphql/resolvers");
+
+const users = require("./data/users");
+
+const server = new ApolloServer({
+  typeDefs,
+  resolvers,
+  context: async ({ req }) => {
```
Comment (davidbarral) [inline index.js:13]: You are handling the tokens, but you are not protecting the queries/mutations using the information that you get via the token.

I can get the operations without using any auth header. Given your implementation it's not clear what you pretend. You use a per-user token but you won't limit the response data using the authenticated user.

Also the `createUser` mutation may need another kind of authorization, something like an admin-level authorization of sorts.

File: graphql/schema.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  }
+
+  type Token {
+    value: String!
+  }
+
+  type Account {
+    number: String!
+    balance: Int!
+    deposit: Int!
```
Comment (davidbarral) [inline graphql/schema.js:21]: Schema seems ok. We are guessing that you use `deposit` and `withdraw` as totalizers. No need for that I think :D

File: graphql/resolvers.js [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+  Query: {
+    getAccountAndMovements: () => operations,
+    me: (root, args, context) => {
+      return context.currentUser;
+    },
+  },
+  Mutation: {
+    addDeposit: (_, args) => {
+      const { number, type, amount } = args;
+      const findNumberAccount = operations
```
Comment (davidbarral) [inline graphql/resolvers.js:18]: Why do you do all this mapping? 🤔.

These resolvers should be simpler, something like:

```
accountsRepo.deposit(accountId, amount);
```

And you should hide the logic behind the data management in that accountsRepo (name is not important, just the idea).

File: graphql/resolvers.js [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        .join();
+      const getActualDeposit = parseInt(
+        operations.map((operation) => operation.deposit)
+      );
+      if (findNumberAccount !== number) {
+        throw new UserInputError("Account number does not match", {
+          invalidArgs: number,
+        });
+      }
+      if (type !== "DEPOSIT") {
```
Comment (davidbarral) [inline graphql/resolvers.js:29]: Why do you need this check?

File: graphql/schema.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  }
+
+  type Query {
+    getAccountAndMovements: [Account!]
+    me: User
+  }
+
+  type Mutation {
+    addDeposit(number: String!, type: MovementTypes!, amount: Int!): Account!
+    makeWithdraw(number: String!, type: MovementTypes!, amount: Int!): Account!
```
Comment (davidbarral) [inline graphql/schema.js:40]:
```suggestion
    addDeposit(number: String!, amount: Int!): Account!
    makeWithdraw(number: String!, amount: Int!): Account!
```

Type is implicit.

File: graphql/resolvers.js [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        .join();
+      const getActualWithdraw = parseInt(
+        operations.map((operation) => operation.withdraw)
+      );
+      if (findNumberAccount !== number) {
+        throw new UserInputError("Account number does not match", {
+          invalidArgs: number,
+        });
+      }
+      if (type !== "WITHDRAW") {
```
Comment (davidbarral) [inline graphql/resolvers.js:73]: The unnecessary check again.

File: graphql/resolvers.js [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        throw new UserInputError("Account number does not match", {
+          invalidArgs: number,
+        });
+      }
+      if (type !== "DEPOSIT") {
+        throw new UserInputError("Only deposit type operations are allowed", {
+          invalidArgs: type,
+        });
+      }
+      if (Math.sign(amount) === -1) {
```
Comment (davidbarral) [inline graphql/resolvers.js:34]: Extracting this to some functions would clear the code.

File: graphql/resolvers.js [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+        throw new UserInputError(
+          "Negative numbers are not allowed in deposit type operations",
+          {
+            invalidArgs: amount,
+          }
+        );
+      }
+      const newBalance =
+        parseInt(operations.map((operation) => operation.balance)) + amount;
```
Comment (davidbarral) [inline graphql/resolvers.js:43]: Why would yo need this `parseInt`?

**Review by davidbarral — CHANGES_REQUESTED** [review]
The resolvers are a bit messy, the code would benefit from more layers and a clearer separation. As it is now, the business and data logic is all mixed and entangled. It's hard to follow.

Can you refactor the code a bit?

Also the authentication/authorization needs more work :)

File: graphql/resolvers.js (hunk 10 above)
Comment (aitorarias) [reply graphql/resolvers.js:43]: To avoid returning a string. 
For instance, "8000" + 20, will return "80020" beceuse of type coercion.

File: graphql/resolvers.js (hunk 9 above)
Comment (aitorarias) [reply graphql/resolvers.js:34]: Absolutely. I'll fix.

File: graphql/resolvers.js (hunk 8 above)
Comment (aitorarias) [reply graphql/resolvers.js:73]: Why unnecessary? That mutation requires pass as a parameter whether it is a deposit or a withdrawal, that is why I check. Don't you think?

File: graphql/schema.js (hunk 7 above)
Comment (aitorarias) [reply graphql/schema.js:40]: My bad. Now I understand.

File: graphql/schema.js (hunk 4 above)
Comment (aitorarias) [reply graphql/schema.js:21]: OK, I'll see how I can do it without being an enum. :)

File: index.js (hunk 3 above)
Comment (aitorarias) [reply index.js:13]: Right. 

Could create roles to protect certain queries or mutations.

**Comment by aitorarias** [comment]
> The resolvers are a bit messy, the code would benefit from more layers and a clearer separation. As it is now, the business and data logic is all mixed and entangled. It's hard to follow.
> 
> Can you refactor the code a bit?
> 
> Also the authentication/authorization needs more work :)

Fully agree with the annotations. 

I have just read them and I am working on it. 

Thank you very much, David :)

File: graphql/resolvers.js (hunk 10 above)
Comment (davidbarral) [reply graphql/resolvers.js:43]: Yeah, that much I understand. My problem is the whole operations mapping 🤔 why parseInt an array?
